File: desktop/macos/Sources/VoiceWorkbenchMac/UpdateRelease.swift
import Foundation

struct UpdateRelease: Decodable {
    struct Asset: Decodable {
        let name: String
        let browser_download_url: URL
    }

    let tag_name: String
    let draft: Bool
    let prerelease: Bool
    let assets: [Asset]

    var version: String { String(tag_name.dropFirst()) }

    func package(newerThan current: String) throws -> Asset? {
        guard !draft, !prerelease else { return nil }
        guard tag_name.hasPrefix("v"), Self.parts(version) != nil,
              Self.parts(current) != nil else { throw Self.failure("更新版本号无效。") }
        guard version.compare(current, options: .numeric) == .orderedDescending else { return nil }
        return try asset(named: "yovoice-\(tag_name)-macos-arm64.zip")
    }

    func asset(named name: String) throws -> Asset {
        let matches = assets.filter { $0.name == name }
        guard matches.count == 1, let asset = matches.first,
              asset.browser_download_url.scheme == "https",
              asset.browser_download_url.host == "github.com",
              asset.browser_download_url.user == nil,
              asset.browser_download_url.password == nil,
              asset.browser_download_url.port == nil,
              asset.browser_download_url.path == "/leemysw/yovoice/releases/download/\(tag_name)/\(name)" else {
            throw Self.failure("当前版本缺少有效的更新包或校验文件，请从 GitHub 下载。")
        }
        return asset
    }

    static func checksum(_ text: String, for name: String) throws -> String {
        let hashes = text.split(whereSeparator: \.isNewline).compactMap { line -> String? in
            let fields = line.split(maxSplits: 1, whereSeparator: \.isWhitespace)
            guard fields.count == 2 else { return nil }
            var filename = String(fields[1]).trimmingCharacters(in: .whitespaces)
            if filename.hasPrefix("*") { filename.removeFirst() }
            if filename.hasPrefix("./") { filename.removeFirst(2) }
            guard filename == name else { return nil }
            return String(fields[0]).lowercased()
        }
        guard hashes.count == 1, let hash = hashes.first,
              hash.count == 64, hash.allSatisfy({ "0123456789abcdef".contains($0) }) else {
            throw failure("更新包的 SHA-256 校验信息无效。")
        }
        return hash
    }

    private static func parts(_ version: String) -> [Int]? {
        let parts = version.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3, parts.allSatisfy({ !$0.isEmpty && $0.allSatisfy(\.isASCII) && $0.allSatisfy(\.isNumber) && Int($0) != nil }) else { return nil }
        return parts.compactMap { Int($0) }
    }

    static func failure(_ message: String) -> NSError {
        NSError(domain: "yovoice.Update", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    }
}

File: desktop/windows/VoiceWorkbench/UpdateRelease.cs
using System.IO;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace VoiceWorkbench.Desktop;

public sealed record UpdateAsset(string Name, [property: JsonPropertyName("browser_download_url")] Uri Url);
public sealed record UpdateRelease(
    [property: JsonPropertyName("tag_name")] string Tag,
    bool Draft, bool Prerelease, UpdateAsset[] Assets)
{
    public string Version => Tag[1..];
    public UpdateAsset? Package(string current)
    {
        if (Draft || Prerelease) return null;
        if (!Regex.IsMatch(Tag, @"^v[0-9]+\.[0-9]+\.[0-9]+$") ||
            !System.Version.TryParse(Version, out var latest) || !System.Version.TryParse(current, out var running))
            throw new IOException("更新版本号无效。");
        if (latest <= running) return null;
        return Asset($"yovoice-{Tag}-windows-x64-setup.exe");
    }
    public UpdateAsset Asset(string name)
    {
        var matches = Assets.Where(asset => asset.Name == name).ToArray();
        if (matches.Length != 1) throw new IOException("当前版本缺少更新包或校验文件。");
        var asset = matches[0];
        if (!asset.Url.IsAbsoluteUri || asset.Url.Scheme != "https" || asset.Url.Host != "github.com" ||
            !asset.Url.IsDefaultPort || asset.Url.UserInfo.Length != 0 ||
            asset.Url.AbsolutePath != $"/leemysw/yovoice/releases/download/{Tag}/{name}")
            throw new IOException("更新文件地址无效。");
        return asset;
    }
    public static string Checksum(string text, string name)
    {
        var hashes = new List<string>();
        foreach (string line in text.Split('\n'))
        {
            var fields = line.Trim().Split((char[]?)null, 2, StringSplitOptions.RemoveEmptyEntries);
            if (fields.Length != 2) continue;
            string filename = fields[1].TrimStart('*');
            if (filename.StartsWith("./", StringComparison.Ordinal)) filename = filename[2..];
            if (filename == name) hashes.Add(fields[0]);
        }
        if (hashes.Count != 1 || !Regex.IsMatch(hashes[0], "^[0-9a-fA-F]{64}$"))
            throw new IOException("更新包的 SHA-256 校验信息无效。");
        return hashes[0];
    }
}
